pkg/webhook/subnet/validator.go

```diff
@@ -0,0 +1,151 @@
+	vpcCache kubeovnnetworkv1.VpcCache
+}
+
+var _ admission.Validator = &Validator{}
+
+func NewSubnetValidator(nadCache ctlcniv1.NetworkAttachmentDefinitionCache, vpcCache kubeovnnetworkv1.VpcCache) *Validator {
+	return &Validator{
+		nadCache: nadCache,
+		vpcCache: vpcCache,
+	}
+}
+
+func (v *Validator) Create(_ *admission.Request, newObj runtime.Object) error {
+	subnet := newObj.(*kubeovnv1.Subnet)
+
+	subnetSpec := subnet.Spec
+
+	if subnetSpec.Provider == "" {
+		return fmt.Errorf("provider is empty for subnet %s in %s", subnet.Name, subnet.Namespace)
+	}
+
+	if subnetSpec.CIDRBlock == "" || subnetSpec.Gateway == "" {
+		return fmt.Errorf("subnet cidr/gw is empty for subnet %s in %s", subnet.Name, subnet.Namespace)
+	}
+
+	err := v.checkCIDRGW(subnetSpec.CIDRBlock, subnetSpec.Gateway)
+	if err != nil {
+		return fmt.Errorf("invalid cidr/gw for subnet %s in %s err=%v", subnet.Name, subnet.Namespace, err)
+	}
+
+	err = v.checkifNadExists(subnetSpec.Provider)
+	if err != nil {
+		return fmt.Errorf("nad does not exists for subnet %s in %s err=%v", subnet.Name, subnet.Namespace, err)
+	}
+
+	err = v.checkifVpcExists(subnetSpec.Vpc)
+	if err != nil {
+		return fmt.Errorf("vpc does not exists for subnet %s in %s err=%v", subnet.Name, subnet.Namespace, err)
+	}
+
+	return nil
+}
+
+func (v *Validator) Update(_ *admission.Request, oldObj, newObj runtime.Object) error {
+	newSubnet := newObj.(*kubeovnv1.Subnet)
+	oldSubnet := oldObj.(*kubeovnv1.Subnet)
+
+	// ignore the update if the resource is being deleted
+	if newSubnet.DeletionTimestamp != nil {
+		return nil
+	}
+
+	newSubnetSpec := newSubnet.Spec
+	oldSubnetSpec := oldSubnet.Spec
+
+	// skip the update if the config is not changed
+	if reflect.DeepEqual(newSubnetSpec, oldSubnetSpec) {
+		return nil
+	}
+
+	if newSubnetSpec.Provider == "" {
+		return fmt.Errorf("provider is empty for subnet %s in %s", newSubnet.Name, newSubnet.Namespace)
+	}
+
+	err := v.checkCIDRGW(newSubnetSpec.CIDRBlock, newSubnetSpec.Gateway)
+	if err != nil {
+		return fmt.Errorf("invalid cidr/gw for subnet %s in %s err=%v", newSubnet.Name, newSubnet.Namespace, err)
+	}
+
+	err = v.checkifNadExists(newSubnetSpec.Provider)
+	if err != nil {
+		return fmt.Errorf("nad does not exists for subnet %s in %s err=%v", newSubnet.Name, newSubnet.Namespace, err)
+	}
+
+	err = v.checkifVpcExists(newSubnetSpec.Vpc)
+	if err != nil {
+		return fmt.Errorf("vpc does not exists for subnet %s in %s err=%v", newSubnet.Name, newSubnet.Namespace, err)
+	}
+
+	return nil
+}
+
+func (v *Validator) checkifNadExists(subnetProvider string) error {
+	providerName := strings.Split(subnetProvider, ".")
+	_, err := v.nadCache.Get(providerName[1], providerName[0])
+	if err != nil {
+		return fmt.Errorf("failed to get nad %s %s created for the subnet", providerName[1], providerName[0])
+	}
+
+	return nil
+}
+
+func (v *Validator) checkifVpcExists(subnetVpc string) error {
+	if subnetVpc != "" {
+		_, err := v.vpcCache.Get(subnetVpc)
+		if err != nil {
```

Comment:
It would be better if we can separate different error types so that we can provide clear information to the user. For instance, https://github.com/harvester/network-controller-harvester/blob/f9304a1f64914ad989452927a79fe35fba0509bd/pkg/webhook/vlanconfig/validator.go#L186